### DISCUSSION TITLE
How should we update Snowplow schemas in the repository?

### DISCUSSION THREAD

**Discussion by WillDaSilva** [meltano/meltano#6159] [General]
We current store our Snowplow JSON schemas on disk with the following structure:

```
tracking
├── iglu-client-embedded
│   └── schemas
│       └── com.meltano
│           └── schema_name
│               └── jsonschema
│                   └── 1-0-0
├── iglu.json
└── micro.conf
```

When we update any of the schemas, we'll need to bump the version number. Old versions of the schema must continue to be hosted (e.g. by SnowcatCloud), but don't necessarily need to be stored at the head of `main` in the repository. I see two options:

### Option 1: Store all versions in the head of `main`

If we keep all of them at the head of `main` then collecting & uploading them will be easy, e.g. if we ever need to switch away from SnowcatCloud,. The downside is that this will result in clutter, and that will discourage us from updating the schemas.

### Option 2: Store only the most recent version in the head of `main`

We'll still store all version by virtue of the git history. This avoids the clutter on disk, but makes it more difficult to collect all of the old versions of the schemas should we ever need to.

Most of the time, the only reason we need to have the schema in the repository at all is to evolve it, for which we only need the most recent version.

---

With either option we'll have to be careful to ensure we never update a published schema. It would be nice if we could enforce that somehow. Maybe with a git hook, or a GitHub action.

@pnadolny13 @pandemicsyn @aaronsteers

**Comment by pandemicsyn** [comment]
> We current store our Snowplow JSON schemas on disk with the following structure:
> 
> ```
> tracking
> ├── iglu-client-embedded
> │   └── schemas
> │       └── com.meltano
> │           └── schema_name
> │               └── jsonschema
> │                   └── 1-0-0
> ├── iglu.json
> └── micro.conf
> ```
> 
> When we update any of the schemas, we'll need to bump the version number. Old versions of the schema must continue to be hosted (e.g. by SnowcatCloud), but don't necessarily need to be stored at the head of `main` in the repository. I see two options:
> 
> ### Option 1: Store all versions in the head of `main`
> If we keep all of them at the head of `main` then collecting & uploading them will be easy, e.g. if we ever need to switch away from SnowcatCloud,. The downside is that this will result in clutter, and that will discourage us from updating the schemas.
> 
> ### Option 2: Store only the most recent version in the head of `main`
> We'll still store all version by virtue of the git history. This avoids the clutter on disk, but makes it more difficult to collect all of the old versions of the schemas should we ever need to.
> 
> Most of the time, the only reason we need to have the schema in the repository at all is to evolve it, for which we only need the most recent version.
> 
> With either option we'll have to be careful to ensure we never update a published schema. It would be nice if we could enforce that somehow. Maybe with a git hook, or a GitHub action.
> 
> @pnadolny13 @pandemicsyn @aaronsteers

Just like regular DB migrations, we should stow all versions on main. I think snowplow micro requires them anyway.

+1 probably not super urgent but would definitely be nice to have workflow check that only files new to the branch are being modified, or something a long those lines, and flag it for the code owners if an existing one is being updated.

We should also probably have a dedicated set of CODEOWNERS for the schemas as well (with Pat being the primary, since he should IMHO hold absolute veto power 😁 )
